src/cargo/core/compiler/output_sbom.rs

```diff
@@ -0,0 +1,239 @@
+//! cargo-sbom precursor files for external tools to create SBOM files from.
+//! See [`output_sbom`] for more.
+
+use std::collections::BTreeSet;
+use std::io::{BufWriter, Write};
+use std::path::PathBuf;
+
+use cargo_util::paths::{self};
+use cargo_util_schemas::core::PackageIdSpec;
+use itertools::Itertools;
+use serde::Serialize;
+
+use crate::{
+    core::{profiles::Profile, PackageId, Target, TargetKind},
+    util::Rustc,
+    CargoResult,
+};
+
+use super::{unit_graph::UnitDep, BuildRunner, CrateType, Unit};
+
+#[derive(Serialize, Clone, Debug, Copy)]
+#[serde(rename_all = "kebab-case")]
+enum SbomBuildType {
+    /// A package dependency
+    Normal,
+    /// A build script dependency
+    Build,
+}
+
+/// Typed version of a SBOM format version number.
+pub struct SbomFormatVersion<const V: u32>;
+
+impl<const V: u32> Serialize for SbomFormatVersion<V> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_u32(V)
+    }
+}
+
+/// A package dependency
+#[derive(Serialize, Clone, Debug)]
+struct SbomDependency {
+    name: String,
+    package_id: PackageId,
+    version: String,
+    features: Vec<String>,
+}
+
+impl From<&UnitDep> for SbomDependency {
+    fn from(dep: &UnitDep) -> Self {
+        let package_id = dep.unit.pkg.package_id();
+        let name = package_id.name().to_string();
+        let version = package_id.version().to_string();
+        let features = dep
+            .unit
+            .features
+            .iter()
+            .map(|f| f.to_string())
+            .collect_vec();
+
+        Self {
+            name,
+            package_id,
+            version,
+            features,
+        }
+    }
+}
+
+#[derive(Serialize, Clone, Debug)]
+struct SbomPackage {
+    package_id: PackageId,
+    package: String,
+    version: String,
+    features: Vec<String>,
+    build_type: SbomBuildType,
+    extern_crate_name: String,
+    dependencies: Vec<SbomDependency>,
+}
+
+impl SbomPackage {
+    pub fn new(
+        dep: &UnitDep,
+        dependencies: Vec<SbomDependency>,
+        build_type: SbomBuildType,
+    ) -> Self {
+        let package_id = dep.unit.pkg.package_id();
```

Comment:
Same as the top level package. Package ID Spec is the only stable representation in Cargo to describe a package version coordinate.

```suggestion
        let package_id = dep.unit.pkg.package_id().to_spec();
```

Reply:
Updated the `package_id`.